<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <title>Document</title>
</head>

<body>

    <script>
//        try {
//            throw new Error('Whoops!');
//        } catch (e) {
//            console.log(e.name + ': ' + e.message);
//        }


//        try {
//            foo.bar();
//        } catch (e) {
//            if (e instanceof EvalError) {
//                console.log(e.name + ': ' + e.message);
//            } else if (e instanceof RangeError) {
//                console.log(e.name + ': ' + e.message);
//            } else if (e instanceof ReferenceError) {
//                console.log(e.name + ':@ ' + e.message);
//            }
//            // ... etc
//        }


        //--CUSTOM ERROR ES6 -----------
        class CustomError extends Error {
            constructor(foo = 'bar', ...params) {
                // Pass remaining arguments (including vendor specific ones) to parent constructor
                super(...params);

                // Maintains proper stack trace for where our error was thrown (only available on V8)
                if (Error.captureStackTrace) {
                    Error.captureStackTrace(this, CustomError);
                }

                // Custom debugging information
                this.foo = foo;
                this.date = new Date();
            }
        }

        try {
            throw new CustomError('baz', 'bazMessage');
        } catch (e) {
            console.log(e.foo); //baz
            console.log(e.message); //bazMessage
            console.log(e.stack); //stacktrace
            console.log(e.date); //date prop
        }
    </script>

</body>

</html>
